llama-cpp-sys-2/build.rs

```diff
@@ -650,22 +646,28 @@ fn compile_vulkan(cx: &mut Build, cxx: &mut Build) -> &'static str {
 
 fn compile_ggml(mut cx: Build) {
     println!("Compiling GGML..");
+    let ggml_src = LLAMA_PATH.join("ggml").join("src");
```

Comment:
it appears the embedded metal library was moved under the `ggml_src` dir as well:
https://github.com/ggerganov/llama.cpp/commit/f3f65429c44bb195a9195bfdc19a30a79709db7b#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R805

Reply:
@brittlewis12 Yep, copying the `ggml_src`/`ggml_include` logic to `compile_vulkan` fixed it!

Reply:
the relevant fixes for mac are here: https://github.com/utilityai/llama-cpp-rs/commit/f8150791733b90cb8254eaa4e44057553e661256

Reply:
<details><summary>example output from simple</summary>

```
cargo run --release --bin simple --features metal -- --n-len=2048 --prompt "<|start_header_id|>user<|end_header_id|>\n\nshare 5 reasons rust is better than c++<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n" local ~/code/hugtug/meta-llama-3.1-8b-instruct.Q4_K_M.gguf
warning: llama-cpp-sys-2@0.1.63: ./llama.cpp/ggml/src/ggml.c:1944:20: warning: unused function 'ggml_vec_dot_f16_unroll' [-Wunused-function]
warning: llama-cpp-sys-2@0.1.63: inline static void ggml_vec_dot_f16_unroll(const int n, const int xs, float * restrict s, void * restrict xv, ggml_fp16_t * restrict y) {
warning: llama-cpp-sys-2@0.1.63:                    ^
warning: llama-cpp-sys-2@0.1.63: ./llama.cpp/ggml/src/ggml.c:2196:20: warning: unused function 'ggml_vec_gelu_f16' [-Wunused-function]
warning: llama-cpp-sys-2@0.1.63: inline static void ggml_vec_gelu_f16(const int n, ggml_fp16_t * y, const ggml_fp16_t * x) {
warning: llama-cpp-sys-2@0.1.63:                    ^
warning: llama-cpp-sys-2@0.1.63: ./llama.cpp/ggml/src/ggml.c:2917:17: warning: unused function 'ggml_get_numa_affinity' [-Wunused-function]
warning: llama-cpp-sys-2@0.1.63: static uint32_t ggml_get_numa_affinity(void) {
warning: llama-cpp-sys-2@0.1.63:                 ^
warning: llama-cpp-sys-2@0.1.63: ./llama.cpp/ggml/src/ggml.c:3308:19: warning: unused function 'ggml_up32' [-Wunused-function]
warning: llama-cpp-sys-2@0.1.63: static inline int ggml_up32(int n) {
warning: llama-cpp-sys-2@0.1.63:                   ^
warning: llama-cpp-sys-2@0.1.63: ./llama.cpp/ggml/src/ggml.c:19482:13: warning: unused function 'ggml_opt_get_grad' [-Wunused-function]
warning: llama-cpp-sys-2@0.1.63: static void ggml_opt_get_grad(int np, struct ggml_tensor * const ps[], float * g) {
warning: llama-cpp-sys-2@0.1.63:             ^
warning: llama-cpp-sys-2@0.1.63: ./llama.cpp/ggml/src/ggml-metal.m:2850:28: warning: unused variable 'error_code' [-Wunused-variable]
warning: llama-cpp-sys-2@0.1.63:                 NSString * error_code = [command_buffer error].localizedDescription;
warning: llama-cpp-sys-2@0.1.63:                            ^
warning: llama-cpp-sys-2@0.1.63: ./llama.cpp/ggml/src/ggml-metal.m:251:13: warning: unused function 'ggml_metal_log' [-Wunused-function]
warning: llama-cpp-sys-2@0.1.63: static void ggml_metal_log(enum ggml_log_level level, const char * format, ...){
warning: llama-cpp-sys-2@0.1.63:             ^
warning: llama-cpp-sys-2@0.1.63: 2 warnings generated.
warning: llama-cpp-sys-2@0.1.63: ./llama.cpp/ggml/src/ggml-alloc.c:467:13: warning: unused function 'ggml_gallocr_set_node_offset' [-Wunused-function]
warning: llama-cpp-sys-2@0.1.63: static void ggml_gallocr_set_node_offset(ggml_gallocr_t galloc, struct ggml_tensor * node, int buffer_id, size_t offset) {
warning: llama-cpp-sys-2@0.1.63:             ^
warning: llama-cpp-sys-2@0.1.63: 1 warning generated.
warning: llama-cpp-sys-2@0.1.63: ./llama.cpp/ggml/src/ggml-quants.c:1767:14: warning: unused function 'make_qkx1_quants' [-Wunused-function]
warning: llama-cpp-sys-2@0.1.63: static float make_qkx1_quants(int n, int nmax, const float * restrict x, uint8_t * restrict L, float * restrict the_min,
warning: llama-cpp-sys-2@0.1.63:              ^
warning: llama-cpp-sys-2@0.1.63: ./llama.cpp/ggml/src/ggml-quants.c:13752:12: warning: unused function 'iq1_find_best_neighbour' [-Wunused-function]
warning: llama-cpp-sys-2@0.1.63: static int iq1_find_best_neighbour(const uint16_t * restrict neighbours, const uint64_t * restrict grid,
warning: llama-cpp-sys-2@0.1.63:            ^
warning: llama-cpp-sys-2@0.1.63: 2 warnings generated.
warning: llama-cpp-sys-2@0.1.63: 5 warnings generated.
    Finished `release` profile [optimized] target(s) in 0.11s
     Running `target/release/simple --n-len=2048 --prompt '<|start_header_id|>user<|end_header_id|>\n\nshare 5 reasons rust is better than c++<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n' local /Users/tito/code/hugtug/meta-llama-3.1-8b-instruct.Q4_K_M.gguf`
llama_model_loader: loaded meta data with 29 key-value pairs and 291 tensors from /Users/tito/code/hugtug/meta-llama-3.1-8b-instruct.Q4_K_M.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = Meta Llama 3.1 8B Instruct
llama_model_loader: - kv   3:                           general.finetune str              = Instruct
llama_model_loader: - kv   4:                           general.basename str              = Meta-Llama-3.1
llama_model_loader: - kv   5:                         general.size_label str              = 8B
llama_model_loader: - kv   6:                            general.license str              = llama3.1
llama_model_loader: - kv   7:                               general.tags arr[str,6]       = ["facebook", "meta", "pytorch", "llam...
llama_model_loader: - kv   8:                          general.languages arr[str,8]       = ["en", "de", "fr", "it", "pt", "hi", ...
llama_model_loader: - kv   9:                          llama.block_count u32              = 32
llama_model_loader: - kv  10:                       llama.context_length u32              = 131072
llama_model_loader: - kv  11:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv  12:                  llama.feed_forward_length u32              = 14336
llama_model_loader: - kv  13:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv  14:              llama.attention.head_count_kv u32              = 8
llama_model_loader: - kv  15:                       llama.rope.freq_base f32              = 500000.000000
llama_model_loader: - kv  16:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  17:                          general.file_type u32              = 15
llama_model_loader: - kv  18:                           llama.vocab_size u32              = 128256
llama_model_loader: - kv  19:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv  20:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  21:                         tokenizer.ggml.pre str              = llama-bpe
llama_model_loader: - kv  22:                      tokenizer.ggml.tokens arr[str,128256]  = ["!", "\"", "#", "$", "%", "&", "'", ...
llama_model_loader: - kv  23:                  tokenizer.ggml.token_type arr[i32,128256]  = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
llama_model_loader: - kv  24:                      tokenizer.ggml.merges arr[str,280147]  = ["Ġ Ġ", "Ġ ĠĠĠ", "ĠĠ ĠĠ", "...
llama_model_loader: - kv  25:                tokenizer.ggml.bos_token_id u32              = 128000
llama_model_loader: - kv  26:                tokenizer.ggml.eos_token_id u32              = 128009
llama_model_loader: - kv  27:                    tokenizer.chat_template str              = {% set loop_messages = messages %}{% ...
llama_model_loader: - kv  28:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type q4_K:  193 tensors
llama_model_loader: - type q6_K:   33 tensors
llm_load_vocab: special tokens cache size = 256
llm_load_vocab: token to piece cache size = 0.7999 MB
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = llama
llm_load_print_meta: vocab type       = BPE
llm_load_print_meta: n_vocab          = 128256
llm_load_print_meta: n_merges         = 280147
llm_load_print_meta: vocab_only       = 0
llm_load_print_meta: n_ctx_train      = 131072
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_layer          = 32
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 8
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_swa            = 0
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 4
llm_load_print_meta: n_embd_k_gqa     = 1024
llm_load_print_meta: n_embd_v_gqa     = 1024
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-05
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 0.0e+00
llm_load_print_meta: n_ff             = 14336
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 0
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 500000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_ctx_orig_yarn  = 131072
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = 8B
llm_load_print_meta: model ftype      = Q4_K - Medium
llm_load_print_meta: model params     = 8.03 B
llm_load_print_meta: model size       = 4.58 GiB (4.89 BPW)
llm_load_print_meta: general.name     = Meta Llama 3.1 8B Instruct
llm_load_print_meta: BOS token        = 128000 '<|begin_of_text|>'
llm_load_print_meta: EOS token        = 128009 '<|eot_id|>'
llm_load_print_meta: LF token         = 128 'Ä'
llm_load_print_meta: EOT token        = 128009 '<|eot_id|>'
llm_load_print_meta: max token length = 256
llm_load_tensors: ggml ctx size =    0.27 MiB
llm_load_tensors: offloading 32 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 33/33 layers to GPU
llm_load_tensors:      Metal buffer size =  4403.50 MiB
llm_load_tensors:        CPU buffer size =   281.81 MiB
.......................................................................................
llama_new_context_with_model: n_ctx      = 2048
llama_new_context_with_model: n_batch    = 2048
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: flash_attn = 0
llama_new_context_with_model: freq_base  = 500000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:      Metal KV buffer size =   256.00 MiB
llama_new_context_with_model: KV self size  =  256.00 MiB, K (f16):  128.00 MiB, V (f16):  128.00 MiB
llama_new_context_with_model:        CPU  output buffer size =     0.49 MiB
llama_new_context_with_model:      Metal compute buffer size =   258.50 MiB
llama_new_context_with_model:        CPU compute buffer size =    12.01 MiB
llama_new_context_with_model: graph nodes  = 1030
llama_new_context_with_model: graph splits = 2
n_len = 2048, n_ctx = 2048, k_kv_req = 2048

<|start_header_id|>user<|end_header_id|>\n\nshare 5 reasons rust is better than c++<|start_header_id|>assistant<|end_header_id|>\n\nHere are five reasons why Rust might be considered better than C++ for certain use cases:

1.  **Memory Safety**: Rust's ownership system and borrow checker ensure that memory is safely managed, eliminating the possibility of null pointer dereferences, data races, and other memory-related bugs that are common in C++.

2.  **Error Handling**: Rust's error handling system is more comprehensive and expressive than C++. Rust's `Result` and `Option` types provide a way to handle errors in a more explicit and safe manner, making it easier to write robust and reliable code.

3.  **Concurrency**: Rust's ownership system and borrow checker also make it easier to write concurrent code that is safe and efficient. Rust's `std::sync` module provides a range of synchronization primitives that are designed to be safe and efficient.

4.  **Performance**: Rust is designed to be a systems programming language, and it is capable of generating highly optimized machine code. Rust's performance is often comparable to C++'s, but with the added safety and convenience of a higher-level language.

5.  **Ecosystem**: Rust's ecosystem is growing rapidly, with a wide range of libraries and frameworks available for tasks such as web development, data science, and machine learning. Rust's package manager, Cargo, makes it easy to manage dependencies and build projects.

It's worth noting that C++ is still a widely used and powerful language, and it may be the better choice for certain use cases, such as systems programming or high-performance computing. However, for many developers, Rust's safety features, error handling, and concurrency support make it a more attractive choice.


decoded 334 tokens in 12.95 s, speed 25.79 t/s

load time = 364.89 ms
sample time = 147.60 ms / 335 runs (0.44 ms per token, 2269.59 tokens per second)
prompt eval time = 295.14 ms / 22 tokens (13.42 ms per token, 74.54 tokens per second)
eval time = 12789.50 ms / 334 runs (38.29 ms per token, 26.12 tokens per second)
total time = 13316.22 ms
```

</details>